app/routes.py

```diff
@@ -1,2 +1,326 @@
+
+    for goal in goals:
+        goals_response.append(goal.convert_to_json())
+
+    return jsonify(goals_response), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["GET"], strict_slashes=False)
+def get_single_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+
+    return make_response({"goal":(saved_goal.convert_to_json())}, 200)
+
+
+@goals_bp.route("/<goal_id>", methods=["PUT"], strict_slashes=False)
+def update_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+    
+    form_data = request.get_json()
+
+    saved_goal.title = form_data["title"]
+
+    db.session.commit()
+
+    return make_response({"goal":(saved_goal.convert_to_json())}, 200)
+
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"], strict_slashes=False)
+def delete_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+
+    db.session.delete(saved_goal)
+    db.session.commit()
+    return {"details": f"Goal {saved_goal.goal_id} \"{saved_goal.title}\" successfully deleted"}, 200
+
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST"], strict_slashes=False)
+def post_tasks_to_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+
+    # again, request.get_json() gets the JSON body from the HTTP request sent by the client and formats it as a Python dictionary
+    # So form_data is a dictionary that will show all the Goal attributes for the goal (with ID goal_id) that we want to assign tasks to 
+    form_data = request.get_json()
+
+    # The tasks attribute in this request is a list of whatever tasks the client is requesting to post to this goal
+    # Here we're just taking that list out of IDs and sticking it in task_ids
+    task_ids = form_data["task_ids"]
+
+    # looping through the list of task ids that the client is requesting to post to this specific goal  
+    for each_task_id in task_ids:
+        # using each task_id to query a Task object that needs to be updated with its corresponding match_goal_id foreign key
+        updated_task = Task.query.get_or_404(each_task_id)
+        # grabbing that goal_id from the goal queried above, and assigning that value to the match_goal_id foreign key of each updated task
+        # then continuing to loop through until each task in the post request has been updated 
+        updated_task.match_goal_id = saved_goal.goal_id
+
+    db.session.commit()
+
+    return make_response({"id": saved_goal.goal_id, "task_ids": task_ids}, 200)
+
+
+@goals_bp.route("/<goal_id>/tasks", methods=["GET"], strict_slashes=False)
+def get_tasks_for_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+    saved_goal_tasks = []
+
+    # filtering the Task query for tasks with a foreign key id that matches the input goal_id
+    tasks = Task.query.filter_by(match_goal_id=goal_id)
+
+    for task in tasks:
```

Comment:
I made the following change in the code and all the tests still pass. I hope that helps with understanding what might have been going wrong that caused whatever you had tried before to not work!
```suggestion
    for task in saved_goal.tasks:
```

Reply:
Thank you for taking the time to look at this again! Not being able to get that approach to work was really annoying me, haha. However - Even though that syntax makes  sense to me,  I'm still having trouble getting the wave 5 and 6 tests to pass, which is kinda what I remember happening when I was writing this (I started off trying to write it the way you show here). I know it's not super urgent - maybe we can just check it out during our next 1:1, I think I'm mostly just curious now. Also - I'm curious how you were able to get the tests to pass given that (line 296 of app/routes.py) my goal instance is saved_goal, not goal - such that I would expect goal.tasks to raise a syntax error?  
![Screen Shot 2021-05-25 at 2 39 50 PM](https://user-images.githubusercontent.com/67732598/119572632-e66b3980-bd67-11eb-9264-16ef6c2bdfea.png)

Reply:
The code snippet I have here does say saved_goal so I'm confused by the question here, but yes! Let's absolutely talk about this more in our next 1:1 if not before! We're scheduled to meet again next week!

Reply:
Yes, that's correct - I'm sorry, it was probably confusing because I wrote the comment there, but was referencing a comment you made on line 296